Shivan Kaul XoT review

wtoorop opened the pull request:
These are nits that came out of reviewing client side XoT from @ShivanKaul and @saradickinson .
I am planning to create a tpkg test for it. For that it is convenient to have a `tls-cert-bundle` like how that exists in Unbound, so I'll create an option for that first.

Commit:
Initial implementation of client side XoT.
- Requires TLS 1.3 support
- Uses a new `tls-auth` clause in the `request-xfr` pattern to specify the authentication domain name to use.

saradickinson committed:
Update configparser.y

Co-authored-by: Wouter Wijngaards <wcawijngaards@users.noreply.github.com>

ShivanKaul committed:
Address review comments

Re-add close(fd) instead of shutdown

ShivanKaul committed:
Merge branch 'master' into xot

Commit:
- remove messages about incomplete writes
- move 2 log messages to after EOF
- clean up some whitespace

Commit:
Fix parsing of optional TLS auth filed in request-xfr pattern so correct error line is reported and confcheck unit test now passes
Add 1 zone to confcheck unit text to include XoT

Commit:
Add a warning if some but not all of the request-xfr patterns use XoT for a given zone

wtoorop committed:
Merge branch 'xot' of https://github.com/ShivanKaul/nsd into ShivanKaul-xot

wtoorop committed:
Nits from code review

wtoorop committed:
tls-cert-bundle option

wtoorop committed:
Not blocking SSL_handshake with XoT client

wtoorop committed:
Fix for uninitialised value

wtoorop committed:
tpkg test for xot

wcawijngaards commented on xfrd-tcp.c:
```diff
@@ -573,12 +701,62 @@ xfrd_tcp_open(struct xfrd_tcp_set* set, struct xfrd_tcp_pipeline* tp,
+
+			} else if(tp->handshake_want == SSL_ERROR_SSL) {
+				if(SSL_get_verify_result(tp->ssl) == X509_V_OK)
+					log_crypto_err("xfrd: TLS handshake failed");
```
```suggestion
			} else if(tp->handshake_want == SSL_ERROR_SSL) {
				log_crypto_err("xfrd: TLS handshake failed");
```
The get_verify_result is only necessary when the handshake has completed successfully? With SSL_ERROR_SSL we know it is wrong and the log_crypto_err function prints the details that are in the error stack.

wtoorop replied:
The `tls_verify_callback()` function also prints an error message if the name doesn't match.
I thought to be smart and prevent double printing of the message by first testing if something else occurred  than a verification problem -- since `SSL_get_verify_result()` results `X509_V_OK` if no verification has taken place yet.
But I noticed with the xot.tdir test that no error is logged when the CA doesn't match. So I guess it is better to have double error reporting than no reporting at all in some error conditions.
Will remove the extra `SSL_get_verify_result(tp->ssl) == X509_V_OK` test

wtoorop replied:
Ah, it was actually a type in the xot test that caused the missing CA not to be printed, but I think it is fine (and safe) to just report the error twise (once during verification, and once when the handshake fails). The fixes are in commits 78eace0 and 9603146

wcawijngaards commented on xfrd-tcp.c:
```diff
@@ -737,7 +988,33 @@ xfrd_tcp_write(struct xfrd_tcp_pipeline* tp, xfrd_zone_type* zone)
+
+			} else if(tp->handshake_want == SSL_ERROR_SSL) {
+				if (SSL_get_verify_result(tp->ssl) == X509_V_OK)
+					log_crypto_err("xfrd: TLS handshake failed");
```
```suggestion
			} else if(tp->handshake_want == SSL_ERROR_SSL) {
				log_crypto_err("xfrd: TLS handshake failed");
```
Same as the other get_verify_result comment?

wtoorop replied:
Yes I'll remove that extra test...

wtoorop committed:
Update checkconf tdir test package

wtoorop committed:
Print SSL errors during handshake

Also if they are about certificate verification failure which might already have been reported before,
because otherwise some certificate verification failures may go unreported (like missing CA).

wtoorop committed:
typo in xot.tdir

wcawijngaards reviewed (approved):
The code looks good to merge. The nsd-control ports in the tdir test are unused.

wtoorop committed:
Disable unused control interface with xot.tdir
